### PR TITLE
SQL-2366: implement schema derivation for field setting stages

### PR DESCRIPTION
**Pull request by jchemburkar**

### CODE REVIEW

**Review by pmeredit — APPROVED** [review]
LGTM, great tests!

File: agg-ast/schema_derivation/src/schema_derivation_tests/stage.rs [hunk 1]
```diff
@@ -1141,3 +1141,243 @@ mod union {
         })
     );
 }
+
+mod replace {
+    use super::*;
+
+    test_derive_stage_schema!(
+        replace_root_nested_doc,
```
Comment (nbagnard) [inline agg-ast/schema_derivation/src/schema_derivation_tests/stage.rs:1149]: Can you add a test where replace root is using a field which does not appear in the source schema to make sure things are working as expected.

File: agg-ast/schema_derivation/src/schema_derivation_tests/stage.rs [hunk 2]
```diff
@@ -1141,3 +1141,243 @@ mod union {
...
+                    ..Default::default()
+                }),
+                "foo".to_string() => Schema::Atomic(Atomic::Decimal),
+            },
+            required: set!("foo".to_string(), "name".to_string()),
+            ..Default::default()
+        })
+    );
+
+    test_derive_stage_schema!(
```
Comment (nbagnard) [inline agg-ast/schema_derivation/src/schema_derivation_tests/stage.rs:1206]: Can you add a test where replace_with is using a field which does not appear in the source schema to make sure things are working as expected.

**Review by nbagnard — CHANGES_REQUESTED** [review]
Just a few edge cases that I think are worth testing out.
Otherwise, it looks great :)

File: agg-ast/schema_derivation/src/schema_derivation_tests/stage.rs (hunk 2 above)
Comment (jchemburkar) [reply agg-ast/schema_derivation/src/schema_derivation_tests/stage.rs:1206]: Would the test below cover what you are looking for already? it creates new fields using a separate operator and tests that those are correctly in the replaced schema

Comment (jchemburkar) [reply agg-ast/schema_derivation/src/schema_derivation_tests/stage.rs:1206]: if so this would be the same question for replaceRoot, which also has a test where an expression is being used to replaceRoot rather than simply a field ref

Comment (nbagnard) [reply agg-ast/schema_derivation/src/schema_derivation_tests/stage.rs:1206]: You are right, it covers the same idea and I found the unit test for field_ref_missing, so that should be alright 👍

**Review by nbagnard — APPROVED** [review]
LGTM!
